### PR TITLE
Feature/stop and wait plugin ros2

### PR DESCRIPTION
**Pull request by aswath1**
# PR Details
## Description

stop and wait plugin ros2
## Related Issue

https://github.com/usdot-fhwa-stol/carma-platform/issues/1857
## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp [hunk 1]
```diff
@@ -49,125 +47,116 @@ using oss = std::ostringstream;
...
+  if (req->maneuver_plan.maneuvers[req->maneuver_index_to_plan].type != carma_planning_msgs::msg::Maneuver::STOP_AND_WAIT)
   {
     throw std::invalid_argument("StopAndWait plugin asked to plan non STOP_AND_WAIT maneuver");
   }
 
-  lanelet::BasicPoint2d veh_pos(req.vehicle_state.x_pos_global, req.vehicle_state.y_pos_global);
+  lanelet::BasicPoint2d veh_pos(req->vehicle_state.x_pos_global, req->vehicle_state.y_pos_global);
 
-  ROS_DEBUG_STREAM("planning state x:" << req.vehicle_state.x_pos_global << ", y: " << req.vehicle_state.y_pos_global << ", speed: " << req.vehicle_state.longitudinal_vel);
+  RCLCPP_INFO_STREAM(nh_->get_logger(),"planning state x:" << req->vehicle_state.x_pos_global << ", y: " << req->vehicle_state.y_pos_global << ", speed: " << req->vehicle_state.longitudinal_vel);
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:78]: Please have the logging level same. Ditto for all other logging statements

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp [hunk 2]
```diff
@@ -49,125 +47,116 @@ using oss = std::ostringstream;
...
 
-  if (req.vehicle_state.longitudinal_vel < epsilon_)
+  if (req->vehicle_state.longitudinal_vel < epsilon_)
   {
-    ROS_WARN_STREAM("Detected that car is already stopped! Ignoring the request to plan Stop&Wait");
-    ROS_DEBUG_STREAM("Detected that car is already stopped! Ignoring the request to plan Stop&Wait");
+    RCLCPP_INFO_STREAM(nh_->get_logger(),"Detected that car is already stopped! Ignoring the request to plan Stop&Wait");
+    RCLCPP_INFO_STREAM(nh_->get_logger(),"Detected that car is already stopped! Ignoring the request to plan Stop&Wait");
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:83]: remove this line with DEBUG, it is redundant

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp [hunk 3]
```diff
@@ -373,26 +358,25 @@ std::vector<cav_msgs::TrajectoryPlanPoint> StopandWait::compose_trajectory_from_
...
   for (size_t i = 0; i < points.size(); i++)
   {
-    ROS_DEBUG_STREAM("1d: " << downtracks[i] << " t: " << times[i] << " v: " << filtered_speeds[i]);
+    RCLCPP_INFO_STREAM(nh_->get_logger(),"1d: " << downtracks[i] << " t: " << times[i] << " v: " << filtered_speeds[i]);
   }
 
   auto traj = trajectory_from_points_times_orientations(raw_points, times, yaws, start_time);
 
-  while (traj.back().target_time - traj.front().target_time < ros::Duration(config_.minimal_trajectory_duration))
+  while (rclcpp::Time(traj.back().target_time) - rclcpp::Time(traj.front().target_time) < rclcpp::Duration(config_.minimal_trajectory_duration))
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:376]: `config_.minimal_trajectory_duration` is `double` so rclcpp::Duration(double) is in nanoseconds. So it needs to be `config_.minimal_trajectory_duration * 1e9`

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp [hunk 4]
```diff
@@ -373,26 +358,25 @@ std::vector<cav_msgs::TrajectoryPlanPoint> StopandWait::compose_trajectory_from_
...
   {
-    cav_msgs::TrajectoryPlanPoint new_point = traj.back();
-    new_point.target_time = new_point.target_time + ros::Duration(config_.stop_timestep);
+    carma_planning_msgs::msg::TrajectoryPlanPoint new_point = traj.back();
+    new_point.target_time = rclcpp::Time(new_point.target_time) + rclcpp::Duration(config_.stop_timestep);
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:379]: `config_.stop_timestep` is `double` so rclcpp::Duration(double) is in nanoseconds. So it needs to be config_.stop_timestep * 1e9`

Good rule of thumb to check these kind of error is to search your packages folder using your VSCode search every instances of `Duration(` or `Time(` (pay attention to the parenthesis). OR also instances of `rclcpp::Duration ` or `rclcpp::Time ` (also pay attention to the space). And once you found it, edit with `1e9` or other constructors as you see fit.

File: stop_and_wait_plugin/src/stop_and_wait_plugin_node.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    config_ = StopandWaitConfig();
+ 
+    // Declare parameters
+    config_.minimal_trajectory_duration = declare_parameter<std::string>("minimal_trajectory_duration", config_.minimal_trajectory_duration);
+    config_.stop_timestep = declare_parameter<std::string>("stop_timestep", config_.stop_timestep);
+    config_.trajectory_step_size = declare_parameter<std::string>("trajectory_step_size", config_.trajectory_step_size);
+    config_.accel_limit_multiplier = declare_parameter<std::string>("accel_limit_multiplier", config_.accel_limit_multiplier);
+    config_.accel_limit = declare_parameter<std::string>("/vehicle_acceleration_limit", config_.accel_limit);
+    config_.crawl_speed = declare_parameter<std::string>("crawl_speed", config_.crawl_speed);
+    config_.cernterline_sampling_spacing = declare_parameter<std::string>("cernterline_sampling_spacing", config_.cernterline_sampling_spacing);
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin_node.cpp:36]: let's fix this typo while we are at it
`centerline_sampling_spacing`
Please fix it in .hpp and yaml file as well

File: stop_and_wait_plugin/test/test_stopandwait_plugin.cpp [hunk 6]
```diff
@@ -14,13 +14,13 @@
  * the License.
  */
 
-#include "stop_and_wait_plugin.h"
-#include "stop_and_wait_config.h"
-#include <gtest/gtest.h>
-#include <ros/ros.h>
+#include "stop_and_wait_plugin.hpp"
+#include "stop_and_wait_plugin.hpp"
```
Comment (MishkaMN) [inline stop_and_wait_plugin/test/test_stopandwait_plugin.cpp:18]: remove redundancy

File: carma/launch/guidance.launch.py [hunk 7]
```diff
@@ -72,6 +72,9 @@ def generate_launch_description():
 
     plan_delegator_param_file = os.path.join(
         get_package_share_directory('plan_delegator'), 'config/plan_delegator_params.yaml')
+        
+    stop_and_wait_plugin_param_file = os.path.join(
+        get_package_share_directory('stop_and_wait_plugin'), 'config/stop_and_wait_plugin_params.yaml')
```
Comment (MishkaMN) [inline carma/launch/guidance.launch.py:77]: Please refer to the route_and_following_plugin's comment on where to call this. remove from here.

**Comment by aswath1** [comment]
https://github.com/usdot-fhwa-stol/carma-platform/issues/1857

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp [hunk 8]
```diff
@@ -49,125 +47,115 @@ using oss = std::ostringstream;
...
+                                          const std::string& plugin_name,
+                                          const std::string& version_id)
+  : nh_(nh), wm_(wm), config_(config),plugin_name_(plugin_name), version_id_ (version_id)
+{};
+
+bool StopandWait::plan_trajectory_cb(carma_planning_msgs::srv::PlanTrajectory::Request::SharedPtr req, carma_planning_msgs::srv::PlanTrajectory::Response::SharedPtr resp)
 {
 
-  ROS_DEBUG_STREAM("Starting stop&wait planning");
+  RCLCPP_INFO_STREAM(nh_->get_logger(),"Starting stop&wait planning");
```
Comment (MishkaMN) [inline stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:62]: log statements need to change. it applies to all log statements

File: stop_and_wait_plugin/test/test_stopandwait_plugin.cpp [hunk 9]
```diff
@@ -80,62 +82,49 @@ TEST(StopandWait, TestStopandWaitPlanning)
...
-  testing::InitGoogleTest(&argc, argv);
-  ros::Time::init();
-  ROSCONSOLE_AUTOINIT;
-  if( ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug) ) { // Change to Debug to enable debug logs
-    ros::console::notifyLoggerLevelsChanged();
-  }
-  auto res = RUN_ALL_TESTS();
-  
-  return res;
-}
```
Comment (MishkaMN) [inline stop_and_wait_plugin/test/test_stopandwait_plugin.cpp:141]: Please include ROS 2 version of unit test main. Any other ros2 package will have it like plan_delegator, ILC etc

**Review by MishkaMN — CHANGES_REQUESTED** [review]

**Review by MishkaMN — APPROVED** [review]

File: stop_and_wait_plugin/src/stop_and_wait_plugin.cpp (hunk 8 above)
Comment (MishkaMN) [reply stop_and_wait_plugin/src/stop_and_wait_plugin.cpp:62]: log levels have to stay the same

**Review by MishkaMN — CHANGES_REQUESTED** [review]

**Comment by MishkaMN** [comment]
please try to fix code smells while you are at it

**Review by MishkaMN — APPROVED** [review]

**Pull request merged**